src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala

```diff
@@ -261,15 +298,17 @@ object CodeGenYul extends CodeGenerator {
     // helper function for a common calling pattern below. todo: there may be a slicker way to do
     //  this with https://docs.scala-lang.org/tour/mixin-class-composition.html in the future
     //  once all the cases are written and work
-    def binary_call(s: String, retvar: Identifier, e1: Expression, e2: Expression): Seq[YulStatement] = {
-        val e1_id = nextTemp()
-        val e2_id = nextTemp()
-        translateExpr(e1_id, e1) ++ translateExpr(e2_id, e2) ++ store_then_ret(retvar, binary_ap(s, e1_id, e2_id))
-    }
+    def call(s: String, retvar: Identifier, es: Expression*): Seq[YulStatement] = {
+        // for each expression, make a new temporary variable and translate the expression
+        val es_trans: Seq[(Seq[YulStatement], Identifier)] = es.map(e => {
+            val id = nextTemp()
+            (translateExpr(id, e), id)
+        })
 
-    def unary_call(s: String, retvar: Identifier, e: Expression): Seq[YulStatement] = {
-        val e_id = nextTemp()
-        translateExpr(e_id, e) ++ store_then_ret(retvar, unary_ap(s, e_id))
+        // flatten the resultant sequences and do them first, then make the call to the function using the Ids
+        es_trans.map(x => decl_plain(x._2)) ++
+            es_trans.flatMap(x => x._1) :+
+            assign1(retvar, ap(s, es_trans.map(x => x._2): _*))
```

Comment:
`ap` needs a more descriptive name. Is it "apply"?

Reply:
yeah, "ap" as in "function application site"; i'll change it to "apply" for now and see if that ends up being too long-winded.